### Import Libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### Data

In [2]:
data = pd.DataFrame({
    'HoursStudied': [2, 3, 4, 5, 6, 7],
    'HoursWatchingTV': [4, 3, 2, 1, 0, 0],
    'OutdoorActivityTime': [2, 4, 6, 8, 10, 12],
    'HoursListeningToMusic': [2, 3, 4, 1, 5, 0],
    'WaterConsumed': [5, 6, 5, 6, 4, 5],
    'TestScore': [65, 70, 75, 80, 85, 90]
})

In [3]:
data

HoursStudied  HoursWatchingTV  OutdoorActivityTime  HoursListeningToMusic  \
0             2                4                    2                      2   
1             3                3                    4                      3   
2             4                2                    6                      4   
3             5                1                    8                      1   
4             6                0                   10                      5   
5             7                0                   12                      0   

   WaterConsumed  TestScore  
0              5         65  
1              6         70  
2              5         75  
3              6         80  
4              4         85  
5              5         90

### Correlation

In [4]:
def correlation(features, col_a, col_b):
    x, y = features[col_a], features[col_b]
    x_mean, y_mean = np.mean(x), np.mean(y)
    x_std, y_std = np.std(x), np.std(y)
    return round(np.sum((x - x_mean) * (y - y_mean)) / (len(x) * x_std * y_std),3)

In [5]:
other_cols = data.columns[:-1]
corr = []
for col in other_cols:
    corr.append(correlation(data, col, 'TestScore'))
corr = np.array(corr).reshape(1, 5)
corr = pd.DataFrame(corr, columns = other_cols, index = ['TestScore'])
corr

HoursStudied  HoursWatchingTV  OutdoorActivityTime  \
TestScore           1.0           -0.982                  1.0   

           HoursListeningToMusic  WaterConsumed  
TestScore                   -0.2         -0.355

In [6]:
positive_corr = []
negative_corr = []
no_sign_corr = []
for i in range(corr.shape[1]):
    if corr.iloc[0, i] > 0.5:
        positive_corr.append(corr.columns[i] + ' - Test Score' + ' : ' + str(corr.iloc[0, i]))
    elif corr.iloc[0, i] < -0.5:
        negative_corr.append(corr.columns[i] + ' - Test Score' + ' : ' + str(corr.iloc[0, i]))
    else:
        no_sign_corr.append(corr.columns[i] + ' - Test Score' + ' : ' + str(corr.iloc[0, i]))

print("Positive Correlation :\n",*positive_corr, sep='\n', end='\n\n')
print("Negative Correlation :\n",*negative_corr, sep='\n', end='\n\n')
print("No Significant Correlation :\n",*no_sign_corr, sep='\n')

Positive Correlation :

HoursStudied - Test Score : 1.0
OutdoorActivityTime - Test Score : 1.0

Negative Correlation :

HoursWatchingTV - Test Score : -0.982

No Significant Correlation :

HoursListeningToMusic - Test Score : -0.2
WaterConsumed - Test Score : -0.355


<h3>Positive Correlation</h3>

Hours Studied (+1): The direct link between studying and developing a greater comprehension of the subject matter can be the cause of the positive correlation between study time and test results. Spending more time studying helps people learn topics thoroughly, practise problem-solving, and become more familiar with the content, all of which improve test performance.

Outdoor Activity (+1): The association between test scores and outdoor activity is positive, indicating that outdoor activity may help students do better in school. Exercise is proven to have positive effects on cognitive performance, emotional regulation, and general wellbeing, all of which can improve focus and academic results.

<h3>Negative Correlation</h3>

Television viewing time (-0.982): The significant inverse relationship between the amount of time spent watching TV and test scores highlights the potential time and attention losses associated with excessive television viewing when compared to studying and other productive activities. This emphasises how crucial it is to prioritise learning and studying while effectively managing your time and avoiding distractions.

<h3>No significant correlation</h3>

Hours listening to music (-0.2): The moderately favourable connection between the time spent listening to music and test scores (while not statistically significant) suggests that music may have a small positive influence on focus and relaxation during study sessions. Finding a balance that suits one's learning preferences and concentrate requirements is crucial, but it's important to note that individual preferences for studying settings might differ greatly.

Water Consumed (0.355): The moderately favourable association between the amount of water consumed and test results (while not statistically significant) raises the possibility that maintaining sufficient hydration may have a beneficial effect on cognitive performance. It's crucial to stay hydrated to prevent focus problems, weariness, and a loss of mental alertness, all of which can negatively impact exam performance.

These connections shed light on the complex relationships between several variables and academic achievement, emphasising the importance of efficient study techniques, conscientious time management, and general wellbeing.